### DISCUSSION TITLE
For edgetx built-in crsf protocol add use

### DISCUSSION THREAD

**Discussion by xiaobaishangche** [EdgeTX/edgetx#2337] [Ideas]
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

The popular use of elrs now makes the built-in RF modules of some remote controls deprecated. Most of the built-in interfaces can be DIYed into built-in elrs social-level modules. I hope EdgeTX developers can add the built-in crsf protocol, thank you.

### Describe the solution you'd like

Turning on the built-in crsf protocol makes the player DIY still the built-in tuner of his own elrs

**Comment by pfeerick** [comment]
I'm not quite sure what you are asking for here. If you mean can EdgeTX make it so the builtin RF be used as ELRS, then that is well beyond the scope of the EdgeTX project. EdgeTX is not responsible for the RF link, that is the responsibility of the built in (or external) RF module.

If you mean for an internal module to be able to be replaced with an ELRS module, then assuming it is wired up correctly, EdgeTX already supports this.

**Comment by raphaelcoeffic** [comment]
> If you mean for an internal module to be able to be replaced with an ELRS module, then assuming it is wired up correctly, EdgeTX already supports this.

If I may add:
- only full duplex ELRS modules (so either from Radiomaster, Jumper, or DIY),
- only on radios that have a full duplex UART on the internal module interface (ex; legacy radios from FrSky don't have this).

**Reply by 2xWhiskey** [reply]
> * only on radios that have a full duplex UART on the internal module interface (ex; legacy radios from FrSky don't have this).

Pardon me for barging in...  I am trying to adapt a DIY ELRS module to replace the internal RF module on my Frsky Taranis X-Lite Pro.  Does this radio fall into the incapable category you mentioned?
![20220921_012914514_iOS](https://user-images.githubusercontent.com/113867350/191394316-8aac25c1-06d1-41ec-bc8c-d5253bdd252b.jpg)

**Reply by raphaelcoeffic** [reply]
X-lite pro should be just fine.

**Reply by FearCapone** [reply]
Will the x lite non pro work?

**Reply by raphaelcoeffic** [reply]
It should, as far as I can see (`radio/src/targets/taranis/CMakeLists.txt`):
```cmake
elseif(PCB STREQUAL XLITE)
  set(DEFAULT_INTERNAL_MODULE XJT_PXX1 CACHE STRING "Default internal module")
  set(INTERNAL_MODULE_SERIAL YES)
```

The important part is this: `set(INTERNAL_MODULE_SERIAL YES)`

However, `passthrough` flashing will not be available, as on the Pro has the extra pin required (`radio/src/targets/taranis/hal.h`):
```c++
  #if defined(PCBXLITES)
    #define INTMODULE_BOOTCMD_GPIO         GPIOC
    #define INTMODULE_BOOTCMD_GPIO_PIN     GPIO_Pin_8  // PC.08
    #define INIT_INTMODULE_BOOTCMD_PIN()   GPIO_SetBits(INTMODULE_BOOTCMD_GPIO, INTMODULE_BOOTCMD_GPIO_PIN);
  #endif
```

**Reply by FearCapone** [reply]
Okay i can't compile, where to get it? 😁

**Reply by raphaelcoeffic** [reply]
No need to, just select the internal module in the hardware settings (once properly connected).

**Reply by FearCapone** [reply]
Via rx/tx? Cause i only read pin 8

**Reply by 2xWhiskey** [reply]
So, unless the behavior of the S.PORT line can be changed via an EdgeTX alteration, I don't think you can do an ELRS module connected in the internal RF bay of the Taranis X-Lite base model handsets.  According to a comment I found by [Kilrah](https://github.com/opentx/opentx/issues/7385):

"CRSF communicates on the SPORT line, which is common with the internal module. When the internal module is on it forces that line low and corrupts the CRSF stream."

**Reply by FearCapone** [reply]
The internal T Pro Elrs Module ist hooked up via rx & tx Not s.port/s.bus 1 wire and runs Crsf Protokoll.

**Reply by 2xWhiskey** [reply]
Hopefully this from my own project helps.  Try pins 6 and 7.
![15-pin](https://user-images.githubusercontent.com/113867350/192802929-1509ae0a-4aea-4641-85f4-3576696ef6e1.jpg)

**Reply by FearCapone** [reply]
Did you think it will work without any modifications? I can try it after 7.10.bcause vaccation

**Reply by 2xWhiskey** [reply]
The whole concept of what we are doing is a modification.  I can't tell you to be honest.  My last PCB revision came in a week and a half ago and I've spent most of my free time since troubleshooting and getting help from others.  It was only last night that I finally got it to work.  I am leery of using the 5V lines for power because I would hear weird sounds coming from the handset's speaker.  Instead I ended up taking power from Vmain on the main board and using one of the 5V lines as an enable source for a 3.3V step-up step-down regulator.

**Reply by pfeerick** [reply]
The "weird sounds" are most likely due to RF - it's normal even for recent radios to have noise, which is why they often hard-kill the speaker/amp when not actively playing sound. Simply because they didn't want to add the extra power filtering or regulator 😆

**Reply by 2xWhiskey** [reply]
Is it them, or us?  Cause when I used to use the OEM internal RF module, those sounds weren't a thing.

**Reply by pfeerick** [reply]
If it didn't buzz/noisy before, probably us/the mod. ;) However, in the case of some of the radios I have, particularly as they get a bit beefier with the internal ELRS, have noticeable noise in the background while the speaker is active. So probably not enough shielding or power filtering.

**Comment by Johnex** [comment]
And to help out even more, maybe he just didn't find the option for it.

In 2.7 the setting for internal module type was moved. Press SYS button, go to Hardware tab, scroll down to Internal RF. You can change there to internal CRSF, it will let you use Crossfire / ELRS modules internally in the radio, either via official modules (Radiomaster ELRS module for example) or hacked in (you can put a Crossfire Nano TX inside the radio and just wire it in).

**Comment by caixiaohai0** [comment]
> > If you mean for an internal module to be able to be replaced with an ELRS module, then assuming it is wired up correctly, EdgeTX already supports this.
> 
> If I may add:
> 
> * only full duplex ELRS modules (so either from Radiomaster, Jumper, or DIY),
> * only on radios that have a full duplex UART on the internal module interface (ex; legacy radios from FrSky don't have this).

For the old FRSKY transmitter, is there any way to support built-in CRSF or multi-protocol, can we support full-duplex URAT function in the vacant IO port of MCU?

**Comment by raphaelcoeffic** [comment]
@caixiaohai0 theoretically, this is just about defining the proper pins/UART in `hal.h`. One could possibly use the debug UART on some targets.

**Comment by FearCapone** [comment]
It's working as internal Module (Hardware - internal Module Crsf) 
Jumper T Pro Module, default config. 
Rx pin 3, tx pin 1 on 10.0.0.1\hardware.html

![IMG_20221007_091626](https://user-images.githubusercontent.com/36577758/194494292-10224079-a5b9-4403-a282-ff18cca8dcf3.jpg)
